### PR TITLE
Use Nginx + Basic Auth to secure OSS Airbyte

### PR DESCRIPTION
**Pull request by evantahler**
Per our [community promise](https://airbyte.com/blog/improving-security-for-open-source-airbyte-users), we want default installations of Airbyte to be secure.  This PR uses Nginx as a basic proxy to front the Airbyte server and webapp for `docker-compose` installations.  This helps the user learn into a security mindset.

The first time you boot up Airbyte, you'll see this:

Screen Shot 2022-10-06 at 1 00 50 PM

This PR also include the ability to set a custom 401 Error page for when authentication fails

Screen Shot 2022-10-06 at 4 47 37 PM

This will effect all users of Airbyte going forward who use docker-compose to run the application.

## Demo
**[Loom Video](https://www.loom.com/share/8429bf7f1e454584b5ff4fe67c012f24)**

## Main Changes:
* All docker images communicate over a private docker network
* Only the Nginx container can bridge the private and public networks, and has a host port mapping
* We build upon the public `nginx:latest` image to build a password authentication file on boot, sourcing the name and password from the environment

## TODO
- [x] Use Shared Gradle Builds
- [x] Add a way to disable the auth (by setting empty ENV variables)
- [x] Custom 401 Page
- [x] Update docs to share default passwords
- [x] Create docs about changing the Nginx password
- [ ] ~~Do the same thing for K8s (probably another PR)~~
- [x] Tests?

### CODE REVIEW

**Comment by evantahler** [comment]
Convo with @cgardens 
* Check with GTM about friction 
* add easy to disable

File: nginx/Dockerfile
Comment (tuliren) [inline nginx/Dockerfile:9]: Currently the username and password env variables are set in both here and the `docker-compose` file. Will the one in `docker-compose` override them here? If not, to change the password, users will have to update them in both places. Ideally they should only need to do it in one place.

Maybe something like the following:

```
ARG BASIC_AUTH_USERNAME
ARG BASIC_AUTH_PASSWORD

ENV BASIC_AUTH=${BASIC_AUTH}
ENV BASIC_AUTH_PASSWORD=${BASIC_AUTH_PASSWORD}
```

I am not 100% sure this is correct, but the idea is to:
- Set the variables only in `docker-compose`.
- Pass them from `docker-compose` to the image during build time as `build-arg`.

**Review by tuliren — APPROVED** [review]

File: nginx/Dockerfile
Comment (tuliren) [reply nginx/Dockerfile:9]: If the above approach works, it's better to use different names to avoid confusion:

```dockerfile
ARG AIRBYTE_LOGIN_USERNAME
ARG AIRBYTE_LOGIN_PASSWORD

ENV BASIC_AUTH=${AIRBYTE_LOGIN_USERNAME}
ENV BASIC_AUTH_PASSWORD=${AIRBYTE_LOGIN_PASSWORD}
```

The `AIRBYTE_LOGIN_USERNAME` and `AIRBYTE_LOGIN_PASSWORD` are set in `docker-compose`. They are more user friendly, and are the ones that users should change.

Comment (evantahler) [reply nginx/Dockerfile]: I *think* the convention is to use the same name so they can be find/replaced simply.  The reason for defining the ENV in both the Dockerfile and the docker-compose is that that the Dockerfile sets defaults, so the container can be run outside of the docker-compose orchestrator (e.g. in K8s)

Comment (evantahler) [reply nginx/Dockerfile]: Update - you were right! I moved things into the configs/env files, and the Dockerfile itself is now setting empty values.  Now folks can toggle the feature on and off.

**Comment by evantahler** [comment]
@davinchia I don't really know what to do for K8s - I made some guesses... can you let me know what I need to do?

**Comment by davinchia** [comment]
@evantahler to make sure I understand before I answer the Kube question, with this change, how does an OSS user expose Airbyte Docker?

**Comment by evantahler** [comment]
@davinchia I added a Loom video to help explain things - let me know if you have followup questions!

**Comment by evantahler** [comment]
To simplify things, I've removed the half-baked kube updates from this PR so this change can focus entirely on those users deploying Airbyte with Docker Compose.  A followup PR will be needed for adding the proxy to OSS K8s deployments.

**Comment by evantahler** [comment]
+ @Amruta-Ranade - I took a pass at updating the docs in this PR

File: docs/operator-guides/security.md
Comment (Amruta-Ranade) [inline docs/operator-guides/security.md:35]: I think this section talks about access management and not authentication. I'm gonna say we don't need to add the password content here.

File: docs/operator-guides/security.md
Comment (Amruta-Ranade) [inline docs/operator-guides/security.md:41]: "HTTP basic-auth" feels like an eng-specific detail. Would it suffice to say "Update the default username and password"?

**Comment by davinchia** [comment]
@evantahler thanks for the loom! 

Good idea to split out the Kube stuff, can help out with the Kube specific PR.

**Comment by evantahler** [comment]
Now with a bash-powered test suite, with the proper hooks into gradle:

```
./gradlew :airbyte-proxy:test
Building all of Airbyte.
/Users/evan/workspace/airbyte/airbyte/airbyte-integrations/connectors
Type-safe dependency accessors is an incubating feature.

> Configure project :
configuring docker task for airbyte-bootloader
configuring docker task for airbyte-cli
configuring docker task for airbyte-container-orchestrator
configuring docker task for airbyte-cron
configuring docker task for airbyte-proxy
configuring docker task for airbyte-server
configuring docker task for airbyte-temporal
configuring docker task for airbyte-webapp
configuring docker task for airbyte-workers
configuring docker task for init
configuring docker task for db-lib
configuring docker task for reporter

> Task :airbyte-proxy:bashTest
Testing airbyte proxy...
Stopping airbyte-proxy-test-container
Error response from daemon: No such container: airbyte-proxy-test-container
Error: No such container: airbyte-proxy-test-container
Starting airbyte-proxy-test-container
c562186ca0e99aacd746847fb457f15f2762bd185f95339b74a5a8d2667abf04
Testing access without auth
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   384  100   384    0     0  17941      0 --:--:-- --:--:-- --:--:-- 29538
✔️  access without auth blocked
Testing access with auth
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
✔️  access with auth worked
Stopping airbyte-proxy-test-container
airbyte-proxy-test-container
airbyte-proxy-test-container
Starting airbyte-proxy-test-container with updated password
b15be4c4688f76170902360c839a12badf6c75062f0d2f211acba0c816f32d58
Testing access with orignial paassword
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   384  100   384    0     0  12614      0 --:--:-- --:--:-- --:--:-- 24000
✔️  access with original auth blocked
Testing access updated auth
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
✔️  access with updated auth worked
Stopping airbyte-proxy-test-container
airbyte-proxy-test-container
airbyte-proxy-test-container
Starting airbyte-proxy-test-container with no password
2cd44387f32963d23abd3a9a8db0e677bd6f6176a78d9198968b50d714b64289
Testing access without auth
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
✔️  access without auth allowed when configured
Stopping airbyte-proxy-test-container
airbyte-proxy-test-container
airbyte-proxy-test-container
Tests Passed ✅

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.4/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 44s
17 actionable tasks: 1 executed, 16 up-to-date
```

File: docs/operator-guides/security.md
Comment (evantahler) [reply docs/operator-guides/security.md]: Reverted in e537bdd6e25461026d09f648baf5e060701775a7

File: docs/operator-guides/security.md
Comment (evantahler) [reply docs/operator-guides/security.md]: Updated in e537bdd6e25461026d09f648baf5e060701775a7

**Comment by evantahler** [comment]
I've manually published the first versions of this container with 

```
docker buildx build . --platform "linux/amd64,linux/arm64" --tag airbyte/airbyte-proxy:latest --tag airbyte/airbyte-proxy:0.40.13 --push
```

When this PR is merged, users can use it.

File: docs/operator-guides/security.md
Comment (davinchia) [inline docs/operator-guides/security.md:41]:
```suggestion
- Update the default username and password.
```

Comment (davinchia) [reply docs/operator-guides/security.md:41]: Should we add more detail on how to update this?

File: docs/deploying-airbyte/local-deployment.md
Comment (davinchia) [inline docs/deploying-airbyte/local-deployment.md:21]: should this link to the security Airbyte doc so it's obvious this is something we can easily change?

File: docs/deploying-airbyte/local-deployment.md
Comment (davinchia) [inline docs/deploying-airbyte/local-deployment.md:53]: same thought here on linking to the securing airbyte doc

File: settings.gradle
Comment (davinchia) [inline settings.gradle:94]: nit: alphabeticise

File: .bumpversion.cfg
Comment (davinchia) [inline .bumpversion.cfg:19]: nit: alphabeticise.

File: airbyte-proxy/401.html
Comment (davinchia) [inline airbyte-proxy/401.html:11]: nice!

File: airbyte-proxy/Dockerfile
Comment (davinchia) [inline airbyte-proxy/Dockerfile:14]: one thought: 
* should we remove these envs? Just reading the docker file it looks like these are not set today, when actually these are injected via the .env file.
* or is this how we are injecting the default value today? Should we leave a comment here explaining that in most cases these should be overriden?

File: airbyte-proxy/build.gradle
Comment (davinchia) [inline airbyte-proxy/build.gradle:6]: nit: separate the args on individual lines for readability

File: airbyte-proxy/build.gradle
Comment (davinchia) [inline airbyte-proxy/build.gradle:3]: I believe we can actually get rid of this since this project isn't a java project and the main thing we are doing here is building a docker image. Everything else should continue working.

File: airbyte-proxy/test.sh
Comment (davinchia) [inline airbyte-proxy/test.sh:25]: should already have been built?

File: airbyte-proxy/test.sh
Comment (davinchia) [inline airbyte-proxy/test.sh:12]: what is this sleep for?

File: airbyte-proxy/test.sh
Comment (davinchia) [inline airbyte-proxy/test.sh:16]: should we remove this?

File: airbyte-proxy/test.sh
Comment (davinchia) [inline airbyte-proxy/test.sh:28]: is this in case a current Airbyte Proxy container is running?

File: docker-compose.yaml
Comment (davinchia) [inline docker-compose.yaml:143]: for my understanding - why do this?

File: docker-compose-cloud.buildx.yaml
Comment (davinchia) [inline docker-compose-cloud.buildx.yaml:103]: this is for the publish to Cloud action, I don't think we need the proxy for Cloud so we can remove this.

**Review by davinchia — COMMENTED** [review]
Amazing stuff @evantahler !

Compliments:
* the awesome Loom
* the tests
* generally well commented code

One question before I approve - do we want to publish this with the OSS release? If we want to we should add the proxy directory to the [publish_docker.sh script](https://github.com/airbytehq/airbyte/blob/master/tools/bin/publish_docker.sh#L5).

File: airbyte-proxy/nginx-auth.conf
Comment (nicor88) [inline airbyte-proxy/nginx-auth.conf:15]: let's imagine that I want to use a value different from `http://airbyte-webapp:80` e.g. I want to use the proxy in k8s chart (for now I will add the proxy by myself before the chart has it). I guess that this value won't work, see [here](https://github.com/airbytehq/airbyte/blob/master/charts/airbyte-webapp/templates/service.yaml#L5), as the k8s service is called `{{ .Release.Name }}-airbyte-webapp-svc` this means that can be referenced via `http://{{ .Release.Name }}-airbyte-webapp-svc:80.` 

Could we make proxy_pass customizable via an ENV variable??

File: airbyte-proxy/Dockerfile
Comment (evantahler) [reply airbyte-proxy/Dockerfile:14]: I'll remove them!

File: airbyte-proxy/test.sh
Comment (evantahler) [reply airbyte-proxy/test.sh:28]: Yep!

File: docker-compose-cloud.buildx.yaml
Comment (evantahler) [reply docker-compose-cloud.buildx.yaml:103]: Ah ha! I thought this was how we published the containers in general - looks like that's publish_docker.sh

File: airbyte-proxy/nginx-auth.conf
Comment (davinchia) [reply airbyte-proxy/nginx-auth.conf:15]: @nicor88 we will tackle Kube related auth in a follow up PR!

**Comment by evantahler** [comment]
@davinchia and @nicor88 - thanks for the reviews! I think I've addressed them all via https://github.com/airbytehq/airbyte/pull/17694/commits/aeaf4ecb55e49bef4002e81660ea030c10190bb5, including the dynamic Proxy URL

File: tools/bin/acceptance_test.sh
Comment (evantahler) [inline tools/bin/acceptance_test.sh:24]: Note to others: Acceptance tests will now neeed to opt-out of using basic auth

File: airbyte-proxy/nginx-auth.conf
Comment (nicor88) [reply airbyte-proxy/nginx-auth.conf]: Thanks, @davinchia, my comment is not necessarily related to k8s (I gave more context about a specific use case).

Adding a flexible PROXY_PASS will make the proxy more flexible, so this comment should enable the groundwork for running easily in k8s. (to add this proxy in k8s we need other few pieces: deployment/service/ingress).

**Comment by evantahler** [comment]
(relevant) tests are /finally/ green!
@davinchia & @nicor88, can I get a re-review please?  I'm having a hard time knowing what is really a problem with these flaky tests 😿 

**Note** - This PR is to be merged by @jerri-airbyte or @sw-yx when the community team is ready to announce this.  Also, we will need to manually publish the new docker image for the first release:

```bash
VERSION="1.2.3" # where VERSION is whatever version is the latest OSS release.
docker buildx build . --platform "linux/amd64,linux/arm64" --tag airbyte/proxy:latest --tag airbyte/proxy:$VERSION --push
```

**Comment by evantahler** [comment]
Blocked on https://github.com/airbytehq/airbyte/pull/17864 which will fix up the failing acceptance tests due to a test assumption that Temporal was available on //localhost

**Comment by benmoriceau** [comment]
> Blocked on #17864 which will fix up the failing acceptance tests due to a test assumption that Temporal was available on //localhost

For this PR could you try to re-add the port-forward in temporal? The platform workflow team will fix it later on and you can be unstuck.

**Comment by timroes** [comment]
@evantahler is the `airbyte`/`password` the default username or will this auto generated by default if you use the image? From my experience having a default username/password does not add any level of security to a product, since the password is well known malicious people would simply try it.

If this isn't auto generated yet, is there any chance we can make this password generate the first time the container boots up and log it to the console, so users will always have a unique publically unknown password?

**Comment by evantahler** [comment]
@timroes Yep, I agree that a fixed password isn't the best.  I view this PR as a stop-gap measure until we move authentication into the application (this is an infrastructure PR).  Once there, we'll likely require the user to set a custom password of their choosing. 

I don't think a random password (even if we could do it technically) is a good idea at the infrastructure level because the setup instructions would be something like 

```
...
then run `cat .env | grep PASSWORD` to see the random password set for you to log in 
```
Do you think that adds too much friction for first-time users?

**Comment by evantahler** [comment]
TODO: The octavia-CLI needs to use HTTP headers now!

File: docs/deploying-airbyte/local-deployment.md
Comment (sw-yx) [inline docs/deploying-airbyte/local-deployment.md:20]: in our blogpost we initially promised to randomize this password: "Most people don’t change default passwords, so this password will be autogenerated so as to be unique per user.". 

i understand that having default passwords is easier to support/debug though, just wanted to doublecheck that this is a conscious decision

File: docs/operator-guides/security.md
Comment (sw-yx) [inline docs/operator-guides/security.md:41]:
```suggestion
- Update the default username and password in your environment's `.env` file:
    ```
		# Proxy Configuration
		# Set to empty values, e.g. "" to disable basic auth
		BASIC_AUTH_USERNAME=your_new_username_here
		BASIC_AUTH_PASSWORD=your_new_password_here
    ```
```

File: airbyte-proxy/401.html
Comment (sw-yx) [inline airbyte-proxy/401.html:13]:
```suggestion
      This deployment of Airbyte is protected by HTTP Basic Authentication.
      Please refer to the <a href="https://docs.airbyte.com">Airbyte docs</a> to learn more about:
      <ul>
	      <li><a href="https://docs.airbyte.com/quickstart/deploy-airbyte">the default username/password</a></li>
	      <li><a href="https://docs.airbyte.com/operator-guides/security">securing your Airbyte deployment</a> including turning off or changing the default username/password.</li>
      </ul>
```

File: docs/deploying-airbyte/local-deployment.md
Comment (evantahler) [reply docs/deploying-airbyte/local-deployment.md]: Yes - this PR enhances what we have today. In the future we will revisit this.

**Review by sherifnada — COMMENTED** [review]
@evantahler is there a way to do this without adding a new service to the Docker deployment? it feels like the more services we add the more startup cost there is to the user. Was this a consideration here?

**Comment by evantahler** [comment]
> @evantahler is there a way to do this without adding a new service to the Docker deployment? it feels like the more services we add the more startup cost there is to the user. Was this a consideration here?

I fully expect this to be removed in a quarter or so once we get 'real' user-management into the OSS Platform.  This was selected because it was quick, and a tool folks might be familiar with.  The upside of being a new service is that if users don't want it, it's easy to remove as well.

**Comment by evantahler** [comment]
The changes to get `octavia-cli` working with HTTP Basic auth are now merged into this PR.  Can I please get a re-review, and then we can marge this in 🥳!

**Review by sherifnada — APPROVED** [review]

File: .bumpversion.cfg
Comment (davinchia) [inline .bumpversion.cfg:21]: for my understanding, the changeset in this file is
1) sorting (thanks!)
2) adding the airbyte proxy dockerfile

is that right?

File: tools/bin/publish_docker.sh
Comment (davinchia) [inline tools/bin/publish_docker.sh:6]: 100% on the sorting

File: airbyte-proxy/README.md
Comment (davinchia) [inline airbyte-proxy/README.md:3]: Can we add more detail here about how this works?

Two points I'm thinking of:
- intended to work in conjunction with the `airbyte_internal` network defined in the default docker compose file.
- by default forwards requesting coming in on 8000 and 8001 to the `PROXY_PASS_WEB` and `PROXY_PASS_API` accordingly.

Otherwise it's two hops to understand how this fits in with the general docker deploy.

**Review by davinchia — APPROVED** [review]
Thanks @evantahler ! 

Two minor non-blocking comments.

**Comment by pmossman** [comment]
It looks like this PR broke the `Frontend: Run End-to-End Tests` check in CI (it times out after 2 hours, with error `Health API not available yet. Retrying in 10 seconds...` repeating over and over.

@evantahler is this on your radar? Master has been red since this was merged in. I know you mentioned an issue with `status.airbyte.com` and publishing new OSS versions, but I think the Frontend E2E Tests are a different issue that I haven't seen mentioned yet so wanted to flag it here!

**Pull request merged**
